File: traceharbor-resource-canary/traceharbor-resource-canary-android/src/main/java/com/kernelflux/traceharbor/resource/processor/LazyForkAnalyzeProcessor.java
package com.kernelflux.traceharbor.resource.processor;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Build;

import com.kernelflux.traceharbor.resource.MemoryUtil;
import com.kernelflux.traceharbor.resource.analyzer.model.ActivityLeakResult;
import com.kernelflux.traceharbor.resource.analyzer.model.DestroyedActivityInfo;
import com.kernelflux.traceharbor.resource.config.ResourceConfig;
import com.kernelflux.traceharbor.resource.config.SharePluginInfo;
import com.kernelflux.traceharbor.resource.dumper.HprofFileManager;
import com.kernelflux.traceharbor.resource.watcher.ActivityRefWatcher;
import com.kernelflux.traceharbor.util.TraceHarborHandlerThread;
import com.kernelflux.traceharbor.util.TraceHarborLog;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.Queue;
import java.util.concurrent.LinkedBlockingQueue;

/**
 * Difference between this process and {@link ForkAnalyseProcessor} is analyzing HPROF until the
 * screen is off.
 *
 * @author aurorani
 * @since 2021/10/25
 */
public class LazyForkAnalyzeProcessor extends BaseLeakProcessor {

    private static final String TAG = "TraceHarbor.LeakProcessor.LazyForkAnalyze";

    private static class AnalyzeTask {
        private final File hprof;
        private final String referenceActivity;
        private final String referenceKey;
        private final long dumpStart;

        AnalyzeTask(File hprof, String referenceActivity, String referenceKey, long dumpStart) {
            this.hprof = hprof;
            this.referenceActivity = referenceActivity;
            this.referenceKey = referenceKey;
            this.dumpStart = dumpStart;
        }

        public File getHprof() {
            return hprof;
        }

        public String getReferenceActivity() {
            return referenceActivity;
        }

        public String getReferenceKey() {
            return referenceKey;
        }

        public long getDumpStart() {
            return dumpStart;
        }
    }

    private volatile boolean isInBackground = false;

    private final Queue<AnalyzeTask> lazyTasks = new LinkedBlockingQueue<>();

    private final Runnable analyzeProcessTask = new Runnable() {
        @Override
        public void run() {
            TraceHarborLog.v(TAG, "analyze task start. background: " + isInBackground + ", queue empty: " + lazyTasks.isEmpty());
            while (isInBackground) {
                final AnalyzeTask task = lazyTasks.poll();
                if (task == null) {
                    TraceHarborLog.v(TAG, "task queue is cleared");
                    break;
                }
                analyze(task);
            }
            TraceHarborLog.v(TAG, "analyze task complete. background: " + isInBackground + ", queue empty: " + lazyTasks.isEmpty());
        }
    };

    private final BroadcastReceiver receiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if (Intent.ACTION_SCREEN_OFF.equals(intent.getAction())) {
                TraceHarborLog.v(TAG, "action screen off");
                isInBackground = true;
                TraceHarborHandlerThread.getDefaultHandler().post(analyzeProcessTask);
            } else if (Intent.ACTION_SCREEN_ON.equals(intent.getAction())) {
                TraceHarborLog.v(TAG, "action screen on");
                isInBackground = false;
                TraceHarborHandlerThread.getDefaultHandler().removeCallbacks(analyzeProcessTask);
            }
        }
    };

    public LazyForkAnalyzeProcessor(ActivityRefWatcher watcher) {
        super(watcher);
        final IntentFilter filter = new IntentFilter();
        filter.addAction(Intent.ACTION_SCREEN_ON);
        filter.addAction(Intent.ACTION_SCREEN_OFF);
        watcher.getResourcePlugin().getApplication().registerReceiver(receiver, filter);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        getWatcher().getResourcePlugin().getApplication().unregisterReceiver(receiver);
    }

    @Override
    public boolean process(DestroyedActivityInfo destroyedActivityInfo) {
        publishIssue(SharePluginInfo.IssueType.LEAK_FOUND, ResourceConfig.DumpMode.NO_DUMP, destroyedActivityInfo.mActivityName, destroyedActivityInfo.mKey, "no dump", "0");

        if (Build.VERSION.SDK_INT > ResourceConfig.FORK_DUMP_SUPPORTED_API_GUARD) {
            TraceHarborLog.e(TAG, "cannot fork-dump with unsupported API version " + Build.VERSION.SDK_INT);
            publishIssue(
                    SharePluginInfo.IssueType.ERR_UNSUPPORTED_API,
                    ResourceConfig.DumpMode.LAZY_FORK_ANALYZE,
                    destroyedActivityInfo.mActivityName, destroyedActivityInfo.mKey,
                    "Unsupported API", "0");
            return false;
        }

        getWatcher().triggerGc();

        if (dumpAndAnalyse(
                destroyedActivityInfo.mActivityName,
                destroyedActivityInfo.mKey
        )) {
            getWatcher().markPublished(destroyedActivityInfo.mActivityName, false);
            return true;
        }

        return false;
    }

    private boolean dumpAndAnalyse(String activity, String key) {

        final long dumpStart = System.currentTimeMillis();

        File hprof = null;
        try {
            hprof = HprofFileManager.INSTANCE.prepareHprofFile("LFAP", true);
        } catch (FileNotFoundException e) {
            TraceHarborLog.printErrStackTrace(TAG, e, "");
        }

        if (hprof != null) {
            if (!MemoryUtil.dump(hprof.getPath(), 600)) {
                TraceHarborLog.e(TAG, String.format("heap dump for further analyzing activity with key [%s] was failed, just ignore.",
                        key));
                return false;
            }
        }

        if (hprof == null || hprof.length() == 0) {
            publishIssue(
                    SharePluginInfo.IssueType.ERR_FILE_NOT_FOUND,
                    ResourceConfig.DumpMode.LAZY_FORK_ANALYZE,
                    activity, key, "FileNull", "0");
            TraceHarborLog.e(TAG, "cannot create hprof file");
            return false;
        }

        TraceHarborLog.i(TAG, String.format("dump cost=%sms refString=%s path=%s",
                System.currentTimeMillis() - dumpStart, key, hprof.getPath()));

        TraceHarborLog.i(TAG, "dump complete, push task into lazy analyze task queue");

        // Add lazy task.
        lazyTasks.add(new AnalyzeTask(hprof, activity, key, dumpStart));

        return true;
    }

    private void analyze(AnalyzeTask task) {
        try {
            final long analyseStart = System.currentTimeMillis();

            final ActivityLeakResult leaks = analyze(task.getHprof(), task.getReferenceKey());
            TraceHarborLog.i(TAG, String.format("analyze cost=%sms refString=%s",
                    System.currentTimeMillis() - analyseStart, task.getReferenceKey()));

            if (leaks.mLeakFound) {
                final String leakChain = leaks.toString();
                publishIssue(
                        SharePluginInfo.IssueType.LEAK_FOUND,
                        ResourceConfig.DumpMode.LAZY_FORK_ANALYZE,
                        task.getReferenceActivity(), task.getReferenceKey(), leakChain,
                        String.valueOf(System.currentTimeMillis() - task.getDumpStart()));
                TraceHarborLog.i(TAG, leakChain);
            } else {
                TraceHarborLog.e(TAG, "leak not found");
            }

        } catch (OutOfMemoryError error) {
            publishIssue(
                    SharePluginInfo.IssueType.ERR_ANALYSE_OOM,
                    ResourceConfig.DumpMode.LAZY_FORK_ANALYZE,
                    task.getReferenceActivity(), task.getReferenceKey(), "OutOfMemoryError",
                    "0");
            TraceHarborLog.printErrStackTrace(TAG, error.getCause(), "");
        } finally {
            TraceHarborLog.i(TAG, "analyze complete");
            //noinspection ResultOfMethodCallIgnored
            task.getHprof().delete();
        }
    }
}

File: traceharbor-resource-canary/traceharbor-resource-canary-android/src/main/java/com/kernelflux/traceharbor/resource/processor/LazyForkAnalyzeProcessor.kt
package com.kernelflux.traceharbor.resource.processor

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Build
import com.kernelflux.traceharbor.resource.MemoryUtil
import com.kernelflux.traceharbor.resource.analyzer.model.ActivityLeakResult
import com.kernelflux.traceharbor.resource.analyzer.model.DestroyedActivityInfo
import com.kernelflux.traceharbor.resource.config.ResourceConfig
import com.kernelflux.traceharbor.resource.config.SharePluginInfo
import com.kernelflux.traceharbor.resource.dumper.HprofFileManager
import com.kernelflux.traceharbor.resource.watcher.ActivityRefWatcher
import com.kernelflux.traceharbor.util.TraceHarborHandlerThread
import com.kernelflux.traceharbor.util.TraceHarborLog
import java.io.File
import java.io.FileNotFoundException
import java.util.Queue
import java.util.concurrent.LinkedBlockingQueue

class LazyForkAnalyzeProcessor(
    watcher: ActivityRefWatcher,
) : BaseLeakProcessor(watcher) {
    private class AnalyzeTask(
        val hprof: File,
        val referenceActivity: String,
        val referenceKey: String,
        val dumpStart: Long,
    )

    @Volatile
    private var isInBackground = false
    private val lazyTasks: Queue<AnalyzeTask> = LinkedBlockingQueue()

    private val analyzeProcessTask =
        Runnable {
            TraceHarborLog.v(
                TAG,
                "analyze task start. background: $isInBackground, queue empty: ${lazyTasks.isEmpty()}",
            )
            while (isInBackground) {
                val task = lazyTasks.poll()
                if (task == null) {
                    TraceHarborLog.v(TAG, "task queue is cleared")
                    break
                }
                analyze(task)
            }
            TraceHarborLog.v(
                TAG,
                "analyze task complete. background: $isInBackground, queue empty: ${lazyTasks.isEmpty()}",
            )
        }

    private val receiver =
        object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                if (Intent.ACTION_SCREEN_OFF == intent.action) {
                    TraceHarborLog.v(TAG, "action screen off")
                    isInBackground = true
                    TraceHarborHandlerThread.getDefaultHandler().post(analyzeProcessTask)
                } else if (Intent.ACTION_SCREEN_ON == intent.action) {
                    TraceHarborLog.v(TAG, "action screen on")
                    isInBackground = false
                    TraceHarborHandlerThread.getDefaultHandler().removeCallbacks(analyzeProcessTask)
                }
            }
        }

    init {
        val filter = IntentFilter()
        filter.addAction(Intent.ACTION_SCREEN_ON)
        filter.addAction(Intent.ACTION_SCREEN_OFF)
        watcher.getResourcePlugin().application?.registerReceiver(receiver, filter)
    }

    override fun onDestroy() {
        super.onDestroy()
        watcher.getResourcePlugin().application?.unregisterReceiver(receiver)
    }

    override fun process(destroyedActivityInfo: DestroyedActivityInfo): Boolean {
        publishIssue(
            SharePluginInfo.IssueType.LEAK_FOUND,
            ResourceConfig.DumpMode.NO_DUMP,
            destroyedActivityInfo.mActivityName,
            destroyedActivityInfo.mKey,
            "no dump",
            "0",
        )

        if (Build.VERSION.SDK_INT > ResourceConfig.FORK_DUMP_SUPPORTED_API_GUARD) {
            TraceHarborLog.e(TAG, "cannot fork-dump with unsupported API version ${Build.VERSION.SDK_INT}")
            publishIssue(
                SharePluginInfo.IssueType.ERR_UNSUPPORTED_API,
                ResourceConfig.DumpMode.LAZY_FORK_ANALYZE,
                destroyedActivityInfo.mActivityName,
                destroyedActivityInfo.mKey,
                "Unsupported API",
                "0",
            )
            return false
        }

        watcher.triggerGc()
        return if (dumpAndAnalyse(destroyedActivityInfo.mActivityName, destroyedActivityInfo.mKey)) {
            watcher.markPublished(destroyedActivityInfo.mActivityName, false)
            true
        } else {
            false
        }
    }

    private fun dumpAndAnalyse(activity: String, key: String): Boolean {
        val dumpStart = System.currentTimeMillis()
        var hprof: File? = null
        try {
            hprof = HprofFileManager.prepareHprofFile("LFAP", true)
        } catch (e: FileNotFoundException) {
            TraceHarborLog.printErrStackTrace(TAG, e, "")
        }

        if (hprof != null) {
            if (!MemoryUtil.dump(hprof.path, 600)) {
                TraceHarborLog.e(
                    TAG,
                    "heap dump for further analyzing activity with key [%s] was failed, just ignore.",
                    key,
                )
                return false
            }
        }

        if (hprof == null || hprof.length() == 0L) {
            publishIssue(
                SharePluginInfo.IssueType.ERR_FILE_NOT_FOUND,
                ResourceConfig.DumpMode.LAZY_FORK_ANALYZE,
                activity,
                key,
                "FileNull",
                "0",
            )
            TraceHarborLog.e(TAG, "cannot create hprof file")
            return false
        }

        TraceHarborLog.i(
            TAG,
            "dump cost=%sms refString=%s path=%s",
            System.currentTimeMillis() - dumpStart,
            key,
            hprof.path,
        )
        TraceHarborLog.i(TAG, "dump complete, push task into lazy analyze task queue")
        lazyTasks.add(AnalyzeTask(hprof, activity, key, dumpStart))
        return true
    }

    private fun analyze(task: AnalyzeTask) {
        try {
            val analyseStart = System.currentTimeMillis()
            val leaks: ActivityLeakResult = analyze(task.hprof, task.referenceKey)
            TraceHarborLog.i(
                TAG,
                "analyze cost=%sms refString=%s",
                System.currentTimeMillis() - analyseStart,
                task.referenceKey,
            )

            if (leaks.mLeakFound) {
                val leakChain = leaks.toString()
                publishIssue(
                    SharePluginInfo.IssueType.LEAK_FOUND,
                    ResourceConfig.DumpMode.LAZY_FORK_ANALYZE,
                    task.referenceActivity,
                    task.referenceKey,
                    leakChain,
                    (System.currentTimeMillis() - task.dumpStart).toString(),
                )
                TraceHarborLog.i(TAG, leakChain)
            } else {
                TraceHarborLog.e(TAG, "leak not found")
            }
        } catch (error: OutOfMemoryError) {
            publishIssue(
                SharePluginInfo.IssueType.ERR_ANALYSE_OOM,
                ResourceConfig.DumpMode.LAZY_FORK_ANALYZE,
                task.referenceActivity,
                task.referenceKey,
                "OutOfMemoryError",
                "0",
            )
            TraceHarborLog.printErrStackTrace(TAG, error, "")
        } finally {
            TraceHarborLog.i(TAG, "analyze complete")
            task.hprof.delete()
        }
    }

    companion object {
        private const val TAG = "TraceHarbor.LeakProcessor.LazyForkAnalyze"
    }
}

